### DISCUSSION TITLE
Calendar improvement

### DISCUSSION THREAD

**Discussion by nordcart** [directus/directus#6024] [Feature Requests]
Feature request:

1. Ability to change what weekday the week starts with, in Europe it´s monday.
2. Different colors for events, for example: color 1 = meetings, color 2 = tasks
3. Sync with google calendar.
4. Show week-numbers (screenshot)
5. Mark red-days (holidays, for example sundays) (screenshot)

![image](https://user-images.githubusercontent.com/12813484/120627104-01c2fc80-c46c-11eb-8fa1-1c304c99565b.png)

**Comment by benhaynes** [comment]
1. Definitely
2. I think this should be done with a conditional formatting repeater (eg: if field/value = X, show as RED)
3. Seems out of scope for a Layout
4. I've personally never seen this in a calendar, so I'm not sure about it passing our 80/20 rule
5. No reason to include this in the code, I think this is a custom Layout option... or a Calendar Module/Pane that allows showing data  from multiple Collections (eg: including a `holidays` Collection)

**Reply by rijkvanzanten** [reply]
Re (4): Weird! I thought it was pretty common. Might be another thing that's more commonly used in Europe? I have it enabled in my calendar as well!

![CleanShot 2021-06-03 at 11 22 20@2x](https://user-images.githubusercontent.com/9141017/120670082-0f718580-c45e-11eb-8dd5-bdcfe3c25ead.png)

**Reply by benhaynes** [reply]
Interesting. Let's see what others say here and we can see if it gets critical mass!

**Reply by nordcart** [reply]
> Re (4): Weird! I thought it was pretty common. Might be another thing that's more commonly used in Europe? I have it enabled in my calendar as well!
> 
> ![CleanShot 2021-06-03 at 11 22 20@2x](https://user-images.githubusercontent.com/9141017/120670082-0f718580-c45e-11eb-8dd5-bdcfe3c25ead.png)

It's pretty common here i  europe, we talk alot about, we cannot take your appointment at week x but anyday week y is fine.

**Reply by nordcart** [reply]
> 1. Definitely
> 2. I think this should be done with a conditional formatting repeater (eg: if field/value = X, show as RED)
> 3. Seems out of scope for a Layout
> 4. I've personally never seen this in a calendar, so I'm not sure about it passing our 80/20 rule
> 5. No reason to include this in the code, I think this is a custom Layout option... or a Calendar Module/Pane that allows showing data  from multiple Collections (eg: including a `holidays` Collection)

Maybe we could have categories that can be signed to the event?

String - category;
-Meeting, assign (blue)
-Task, assign (green)
-Important, assign (orange)
-holidays, based on locale, assign (red) (i.e. sundays)

**Reply by rijkvanzanten** [reply]
> Maybe we could have categories that can be signed to the event?

I don't know if that should be on the layout level, or if we should have a "select the color field" dropdown, so the color is controlled by the data itself (instead of the layout)

For category, that could be relational

**Comment by rijkvanzanten** [comment]
@nordcart Wanna open a PR to add an option to toggle "Week starts on:" and "Show Week numbers"? I believe `fullcalendar` natively supports those already, so it should be pretty doable to add that into https://github.com/directus/directus/blob/main/app/src/layouts/calendar/calendar.vue

**Reply by nickrum** [reply]
I think `fullcalendar` determines the "week starts on" based on the locale. Maybe we should set `fullcalendar`'s locale based on the app's locale.

**Reply by nordcart** [reply]
Sure

**Reply by nordcart** [reply]
How do I open a PR? Sorry, I´m new to GitHub 😄

**Reply by rijkvanzanten** [reply]
No worries! This is a very good introduction in the work flow on GitHub 🙂 
https://github.com/firstcontributions/first-contributions

**Reply by nordcart** [reply]
Sorry I was not able to.. Don´t know why

**Comment by jrvidotti** [comment]
Hi! There is any progress in this area, specifically in changing color conditionaly?

**Reply by benhaynes** [reply]
Eventually, but not in the short term without it being sponsored. Busy times!

**Comment by cgm164** [comment]
I would be quite interested in working on a PR for the additional colors feature. 

My idea, would be to add a new option to the calendar layout called "Conditions" and then a list of items and a button to add new conditions. When you click on any item or add a new item, a side panel will open with a form filling in the following data: 
 - Title of the condition
 - Rule
 - Hide (I think this would be interesting too while we're at it)
 - Color
It would be a functionality very similar to the one used in the "Conditions" tab of the fields of the data models.

We can discuss what would be the best way to implement this functionality and once defined I will work on a PR with pleasure :smile:

**Reply by benhaynes** [reply]
Thank you for the offer, we would love the help (once we coordinate how best to get this done).

Currently, Layouts only have user options... which are configured in the sidebar (pretty small) and while you can use presets for roles/global... they are usually set per user. This works for simple options, but I think what you're describing sounds more like "settings" that would exist for the collection+layout for everyone (at least as a default).

I've been talking with @rijkvanzanten about adding the ability for extensions to register settings (configured by admins within the Settings module). This might be a good time to look into how that would operate!

**Reply by rijkvanzanten** [reply]
I'm wondering if this is configuration on the layout, or if we should/can rely on the existing displays for this 🤔

**Comment by hugopvl** [comment]
relatively the number (4) here in europe is used a lot, usually in production companies, footwear, clothing, etc etc. but it is necessary to take into account, as in sqlserver between week and iso_week, they have different start dates, so weeks start on different days. if you compare my print with rijick's print, week 1 of it corresponds to my week 53 of the year 2020.

![image](https://user-images.githubusercontent.com/81326296/174633212-0074ece8-b622-4a99-b094-68bf89638f86.png)

**Reply by rijkvanzanten** [reply]
> if you compare my print with rijick's print, week 1 of it corresponds to my week 53 of the year 2020.

Huh.. That's odd! I'm curious what locale setting in macOS controls that. I don't remember it ever showing week 53 on my end (in either USA or NL locales 🤔 )

**Comment by igiornotigi** [comment]
Hi, I think it would be very helpful if you could insert more start and end fields. For example, if within an item I have 4 date fields, phase-A start phase-A end + phase-B start phase-B end ... Would that be a complicated thing to do?

**Reply by rijkvanzanten** [reply]
> Would that be a complicated thing to do?

Depends! We're currently relying on fullcalendar to render out the calendar, so if they already support such a UX, it should be pretty doable to get that going from a Directus perspective. If they don't however, this becomes way more complicated to support. The other question too is if it's common enough of a use case where it makes sense to include it in the core feature set. Do you have a specific example of how you'd use a date setup like that?

**Comment by igiornotigi** [comment]
I understand, I'll search in fullcalendar to see if there is something similar.
My situation is this: architecture firm, I have a "projects" collection and each project is divided into several stages. Each stage has a start and end date. For example "concept stage" begins on ... ends on ..., "preliminary stage" begins on ... ends on ..., "executive stage" begins on.. ends on ...
So since each phase has two dates, in the example I would have three start dates and three end dates.
Now I have to make a calendar bookmark for each stage, but I can't see two different stages in the same one...
I hope I explained myself 😅

**Reply by rijkvanzanten** [reply]
What do you mean by this? (emphasis mine):

> Now I have to make a calendar bookmark for each stage, **but I can't see two different stages in the same one...**

Thanks for the additional info!

**Comment by igiornotigi** [comment]
I mean that in the "Projects" collection I have to make three bookmarks with calendar layout, one for each stage.
One with start date = "concept stage start date" and end date = "concept stage end date".
Similarly the second will have start date = "preliminary stage start date" and end date = "preliminary stage end date", while the third will have start date = "executive stage start date" and end date = "executive stage end date".
Since the calendars are divided into stages, I cannot see if the concept stage of a project overlaps the executive stage of another, for example.

> Thanks for the additional info!

thank you all for the fantastic work! 🙏

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
